### ISSUE TITLE
Error using Django LDAP Auth with Weblate 

### ISSUE THREAD

**Issue by ymmd** [WeblateOrg/weblate#1460]
Hi,

we have set up Weblate from the repository. We want to enable LDAP login and did all steps as described here: https://docs.weblate.org/en/latest/admin/auth.html#ldap-authentication

This is an excerpt of our settings.py:

```
AUTH_LDAP_SERVER_URI = "ldaps://example.org:636"
AUTH_LDAP_USER_DN_TEMPLATE = "uid=%(user)s,ou=people,dc=example,dc=org"

AUTHENTICATION_BACKENDS = (
    'weblate.accounts.auth.EmailAuth',
    'weblate.accounts.auth.WeblateUserBackend',
    'django_auth_ldap.backend.LDAPBackend'
)
```

The settings for logging are the same as in the settings example file: https://github.com/WeblateOrg/weblate/blob/master/weblate/settings_example.py

When I attempt to login with the right LDAP credentials, everything works. When I provide wrong credentials, I see a full page error message with debug information:

```
Request Method: POST
Request URL: https://example.com/accounts/login/

Django Version: 1.11
Python Version: 2.7.9
Installed Applications:
(u'django.contrib.auth',
 u'django.contrib.contenttypes',
 u'django.contrib.sessions',
 u'django.contrib.sites',
 u'django.contrib.messages',
 u'django.contrib.staticfiles',
 u'django.contrib.admin',
 u'django.contrib.admindocs',
 u'django.contrib.sitemaps',
 u'social_django',
 u'crispy_forms',
 u'compressor',
 u'rest_framework',
 u'rest_framework.authtoken',
 u'weblate.trans',
 u'weblate.lang',
 u'weblate.permissions',
 u'weblate.screenshots',
 u'weblate.accounts',
 u'weblate.utils',
 u'weblate')
Installed Middleware:
(u'django.contrib.sessions.middleware.SessionMiddleware',
 u'django.middleware.common.CommonMiddleware',
 u'django.middleware.locale.LocaleMiddleware',
 u'django.middleware.csrf.CsrfViewMiddleware',
 u'weblate.accounts.middleware.AuthenticationMiddleware',
 u'django.contrib.messages.middleware.MessageMiddleware',
 u'django.middleware.clickjacking.XFrameOptionsMiddleware',
 u'social_django.middleware.SocialAuthExceptionMiddleware',
 u'weblate.accounts.middleware.RequireLoginMiddleware',
 u'weblate.middleware.SecurityMiddleware')



Traceback:

File "/var/www/weblate/env/local/lib/python2.7/site-packages/django/core/handlers/exception.py" in inner
  41.             response = get_response(request)

File "/var/www/weblate/env/local/lib/python2.7/site-packages/django/core/handlers/base.py" in _legacy_get_response
  249.             response = self._get_response(request)

File "/var/www/weblate/env/local/lib/python2.7/site-packages/django/core/handlers/base.py" in _get_response
  187.                 response = self.process_exception_by_middleware(e, request)

File "/var/www/weblate/env/local/lib/python2.7/site-packages/django/core/handlers/base.py" in _get_response
  185.                 response = wrapped_callback(request, *callback_args, **callback_kwargs)

File "/var/www/weblate/weblate/accounts/views.py" in weblate_login
  407.             'title': _('Login'),

File "/var/www/weblate/env/local/lib/python2.7/site-packages/django/contrib/auth/views.py" in inner
  54.         return func(*args, **kwargs)

File "/var/www/weblate/env/local/lib/python2.7/site-packages/django/contrib/auth/views.py" in login
  139.     return LoginView.as_view(**kwargs)(request, *args, **kwargs)

File "/var/www/weblate/env/local/lib/python2.7/site-packages/django/views/generic/base.py" in view
  68.             return self.dispatch(request, *args, **kwargs)

File "/var/www/weblate/env/local/lib/python2.7/site-packages/django/utils/decorators.py" in _wrapper
  67.             return bound_func(*args, **kwargs)

File "/var/www/weblate/env/local/lib/python2.7/site-packages/django/views/decorators/debug.py" in sensitive_post_parameters_wrapper
  76.             return view(request, *args, **kwargs)

File "/var/www/weblate/env/local/lib/python2.7/site-packages/django/utils/decorators.py" in bound_func
  63.                 return func.__get__(self, type(self))(*args2, **kwargs2)

File "/var/www/weblate/env/local/lib/python2.7/site-packages/django/utils/decorators.py" in _wrapper
  67.             return bound_func(*args, **kwargs)

File "/var/www/weblate/env/local/lib/python2.7/site-packages/django/utils/decorators.py" in _wrapped_view
  149.                     response = view_func(request, *args, **kwargs)

File "/var/www/weblate/env/local/lib/python2.7/site-packages/django/utils/decorators.py" in bound_func
  63.                 return func.__get__(self, type(self))(*args2, **kwargs2)

File "/var/www/weblate/env/local/lib/python2.7/site-packages/django/utils/decorators.py" in _wrapper
  67.             return bound_func(*args, **kwargs)

File "/var/www/weblate/env/local/lib/python2.7/site-packages/django/views/decorators/cache.py" in _wrapped_view_func
  57.         response = view_func(request, *args, **kwargs)

File "/var/www/weblate/env/local/lib/python2.7/site-packages/django/utils/decorators.py" in bound_func
  63.                 return func.__get__(self, type(self))(*args2, **kwargs2)

File "/var/www/weblate/env/local/lib/python2.7/site-packages/django/contrib/auth/views.py" in dispatch
  90.         return super(LoginView, self).dispatch(request, *args, **kwargs)

File "/var/www/weblate/env/local/lib/python2.7/site-packages/django/views/generic/base.py" in dispatch
  88.         return handler(request, *args, **kwargs)

File "/var/www/weblate/env/local/lib/python2.7/site-packages/django/views/generic/edit.py" in post
  182.         if form.is_valid():

File "/var/www/weblate/env/local/lib/python2.7/site-packages/django/forms/forms.py" in is_valid
  183.         return self.is_bound and not self.errors

File "/var/www/weblate/env/local/lib/python2.7/site-packages/django/forms/forms.py" in errors
  175.             self.full_clean()

File "/var/www/weblate/env/local/lib/python2.7/site-packages/django/forms/forms.py" in full_clean
  385.         self._clean_form()

File "/var/www/weblate/env/local/lib/python2.7/site-packages/django/forms/forms.py" in _clean_form
  412.             cleaned_data = self.clean()

File "/var/www/weblate/weblate/accounts/forms.py" in clean
  436.                 password=password

File "/var/www/weblate/env/local/lib/python2.7/site-packages/django/contrib/auth/__init__.py" in authenticate
  100.             user = backend.authenticate(*args, **credentials)

File "/var/www/weblate/env/local/lib/python2.7/site-packages/django_auth_ldap/backend.py" in authenticate
  169.             user = ldap_user.authenticate(password)

File "/var/www/weblate/env/local/lib/python2.7/site-packages/django_auth_ldap/backend.py" in authenticate
  342.             self._authenticate_user_dn(password)

File "/var/www/weblate/env/local/lib/python2.7/site-packages/django_auth_ldap/backend.py" in _authenticate_user_dn
  458.             self._bind_as(self.dn, password, sticky=sticky)

File "/var/www/weblate/env/local/lib/python2.7/site-packages/django_auth_ldap/backend.py" in _bind_as
  735.                                              force_str(bind_password))

File "/var/www/weblate/env/local/lib/python2.7/site-packages/ldap/ldapobject.py" in simple_bind_s
  226.     resp_type, resp_data, resp_msgid, resp_ctrls = self.result3(msgid,all=1,timeout=self.timeout)

File "/var/www/weblate/env/local/lib/python2.7/site-packages/ldap/ldapobject.py" in result3
  516.       resp_ctrl_classes=resp_ctrl_classes

File "/var/www/weblate/env/local/lib/python2.7/site-packages/ldap/ldapobject.py" in result4
  523.     ldap_result = self._ldap_call(self._l.result4,msgid,all,timeout,add_ctrls,add_intermediates,add_extop)

File "/var/www/weblate/env/local/lib/python2.7/site-packages/ldap/ldapobject.py" in _ldap_call
  116.         e.args[0]['info'] = strerror(e.args[0]['errno'])

Exception Type: KeyError at /accounts/login/
Exception Value: 'errno'
```

Here are the local vars of the last stack trace item:

```
args: (1, 1, -1, 0, 0, 0)
diagnostic_message_success: None
e: INVALID_CREDENTIALS({'desc': 'Invalid credentials'},)
func: <built-in method result4 of LDAP object at 0x7f57d38ba350>
kwargs: {}
self: <ldap.ldapobject.SimpleLDAPObject instance at 0x7f57d37f0638>
```

Even when I disable debug mode, I get a server error when attempting a login with invalid credentials.

Any ideas how we could prevent django_ldap_auth's exceptions from causing such errors? All we want is that the user sees the usual notice that he provided wrong credentials.

Thank you very much in advance!

Sebastian

**Comment by nijel** [comment]
It seems that python-ldap has some problems with your server. Are you using latest version?

**Comment by ymmd** [comment]
Thanks for your immediate response! I have installed django-auth-ldap (1.2.11) and python-ldap (2.4.35) via pip and just verified that they are both at the latest version.

By now I have tested both, a configuration using Search/bind and a configuration using Direct bind (as described [here](https://pythonhosted.org/django-auth-ldap/authentication.html)). But in both cases there is the same server error, as soon as I enter wrong credentials. (When I enter valid credentials, everything works as expected.)

My understanding is, that django-auth-ldap will throw an error, whenever someone attemts to authenticate with invalid credentials. As soon as there is an exception, Django prints it to the user, I guess. Btw, when debug mode is turned off, the server still sends a 500 with the error message: Server Error The server had serious problems while serving your request. We've just sent our trained monkeys to fix the issue.

Could it be that this is a direct result of some logging mechanism that tries to catch all debug messages? We are using the logging configuration from the [example settings](https://github.com/WeblateOrg/weblate/blob/master/weblate/settings_example.py). I am just reading further into Django logging and hope I'll be able to fix it. We are still grateful for any help.

Sebastian
